packages/client/src/pages/Game/towers/Tower.tsx

```diff
@@ -0,0 +1,243 @@
+
+  public drawTowerRange() {
+    this.engine.context?.beginPath();
+    this.engine.context!.lineWidth = 0.5;
+    this.engine.context?.setLineDash([10, 15]);
+    this.engine.context!.strokeStyle = this.towerParams.strokeStyle;
+    // draw tower range
+    this.engine.context?.arc(
+      this.currentPosition.x - this.towerParams.width / 2,
+      this.currentPosition.y - this.towerParams.height / 2,
+      this.towerParams.attackRange,
+      0,
+      360,
+    );
+    this.engine.context?.stroke();
+    this.engine.context?.closePath();
+  }
+
+  public drawTower() {
+    // draw tower 2d representation
+    this.draw();
+    // draw tower range
+    if (this.engine.isCanBuild) {
+      this.drawTowerRange();
+    }
+  }
+
+  public isEnemyInRange(enemy: Enemy) {
+    const distance =
+      (enemy.currentPosition.x -
+        this.currentPosition.x +
+        this.towerParams.width) *
+        (enemy.currentPosition.x -
+          this.currentPosition.x +
+          this.towerParams.width) +
+      (enemy.currentPosition.y -
+        this.currentPosition.y +
+        this.towerParams.height) *
+        (enemy.currentPosition.y -
+          this.currentPosition.y +
+          this.towerParams.height);
+    if (
+      distance <
+      this.towerParams.attackRange * this.towerParams.attackRange
+    ) {
+      this.target = enemy;
+      return true;
+    }
+    return false;
+  }
+
+  public findTarget() {
+    if (this.engine.enemies!.length) {
+      if (!this.target) {
+        this.engine.enemies?.forEach((enemy) => {
+          return this.isEnemyInRange(enemy);
+        });
+      } else {
+        if (!this.isEnemyInRange(this.target)) {
+          this.target = null;
+        }
+      }
+    }
+  }
+
+  public findTargetVector() {
+    // there is no spoon, Neo
+    if (!this.target) {
+      return;
+    }
+    const xDistance =
+      this.target.currentPosition.x -
+      this.currentPosition.x +
+      this.target.enemyParams.rectCenterX!;
+    const yDistance =
+      this.target.currentPosition.y -
+      this.currentPosition.y +
+      this.target.enemyParams.rectCenterY!;
+    this.towerParams.firingAngle = Math.atan2(yDistance, xDistance) + Math.PI; //* 180 / Math.PI
+  }
+
+  public fire() {
+    const fireFromCoordinates = {
+      x: this.currentPosition.x - this.towerParams.rectCenterX,
+      y: this.currentPosition.y - this.towerParams.rectCenterY,
+    };
+    if (this.isCanFire && this.target) {
+      this.engine.projectiles = [
+        ...this.engine.projectiles!,
+        new Projectile(
+          this.IProjectilemage!,
+          this.target!,
+          this,
+          this.towerParams.attackDamage,
+          fireFromCoordinates,
+        ),
+      ];
+
+      this.isCanFire = false;
+    }
+  }
+
+  public draftBuildTower() {
+    if (this.engine.isCanBuild) {
+      this.draw();
+      this.drawTowerRange();
+    }
+  }
+
+  public destroy() {
+    this.engine.towers!.filter((tower) => this !== tower);
```

Comment:
Here, you return a new array to nowhere. It seems, that the return statement or assignment is missing here. Please, double check

Reply:
Fixed to
`this.engine.towers = this.engine.towers!.filter((tower) => this !== tower);`